pkg/endpoint/policy.go

```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
 // however, and it is possible that policy update succeeds for some endpoints,
 // while it fails for other endpoints.
 //
-// Returns:
-//   - err: any error in obtaining information for computing policy, or if
+// Failure may be due to any error in obtaining information for computing policy,
+// or if policy could not be generated given the current set of rules in the repository.
 //
-// policy could not be generated given the current set of rules in the
-// repository.
-// Must be called with endpoint mutex held.
-func (e *Endpoint) regeneratePolicy() (retErr error) {
-	var forceRegeneration bool
+// endpoint lock must NOT be held. This is because the ipcache needs to be able to
+// make progress while generating policy, and *that* needs the endpoint unlocked to call
+// ep.ApplyPolicyMapChanges. Specifically, computing policy may cause identity allocation
+// which requires ipcache progress.
+//
+// buildMutex MUST be held, and not released until setDesiredPolicy and
+// updateRealizedState have been called
+//
+// There are a few fields that depend on this exact configuration of locking:
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
```

Comment:
Incidentally do you have a particular interaction in mind that may trigger policy drops in such a scenario? Is this another ipcache incremental policy update logic synchronization issue?

Reply:
I don't think this is an issue, just a potential inconsistency that `ep.mutex` has always protected, and I was making explicit.

If we didn't have `ep.mutex`, we could get in the situation where `ep.desiredPolicy` has been updated, but we've not yet called `ep.syncPolicyMap()`. When the ipcache calls `ep.ApplyPolicyMapChanges`, it consumes all deltas queued in `ep.desiredPolicy`. _If we didn't have the endpoint lock_, something like this could happen.

0. `ep.realizedPolicy` selects Identity 1
1. Identity 1 is replaced by 2. (This is the part where I'm fuzzy, but it seems like the selectorcache gets input from both the ipcache and the `identityWatcher`). The selectorcache gets `UpdateIdentities(added=2, deleted=1)`
2. An endpoint recomputes policy. The policy selects only endpoint 2.
3. The new policy is written to `ep.desiredPolicy` **but not synced yet**
4. The ipcache calls `UpdateIdentities(added=2, deleted=[])`
5. The ipcache calls `ep.ApplyPolicyMapChanges`, which is a no-op
6. The ipcache swaps the identity for pods from 1 to 2. At this point, there could be a policy drop, since 2 is not in BPF.
7. The endpoint calls `ep.syncPolicyMap()` which corrects the drop.

At this point we have a policy drop, since we never added identity 2 to the BPF map. Again, **this was never a bug**, since we lock before step 3, ensuring that steps 3 and 7 must happen before steps 5 and 6. I just documented this requirement, since there were concerns that incremental updates could be missed.

Reply:
I wish I had a suggestion on how we might encode this into the commit message or commit content since it seems like a super helpful example, but I'm not quite sure the best format to cover this, other than just having the comment we already have here. :sweat_smile: 

I guess we might consider adding something like ".... policy drops may occur due to the lack of synchronization between full policy generation and incremental policy generation".

Reply:
I added some more words, hopefully those are useful to whomever next stumbles down this path.

Reply:
Did those land in the current version? The changes are not entirely obvious to me. Though I'm also fine with this in general :+1:

Reply:
Ah I see they ended up at `setDesiredPolicy()` :+1: